doc/musig-spec.mediawiki

```diff
@@ -0,0 +1,222 @@
+* Let ''L = HashKeys(pk<sub>1..u</sub>)''.
+* Return 1 if ''IsSecond(pk<sub>1..u</sub>, pk<sub>i</sub>)'', otherwise return  ''int(hash<sub>KeyAgg coefficient</sub>(L || pk) mod n''.
+
+==== Nonce Generation ====
+
+The algorithm ''NonceGen()'' is defined as:
+* Generate two random integers ''k<sub>1</sub>, k<sub>2</sub>'' in the range ''1...n-1''
+* Let ''R<sup>*</sup><sub>1</sub> = k<sub>1</sub>⋅G, R<sup>*</sup><sub>2</sub> = k<sub>2</sub>⋅G''
+* Let ''pubnonce = cbytes(R<sup>*</sup><sub>1</sub>) || cbytes(R<sup>*</sup><sub>2</sub>)''
+* Let ''secnonce = bytes(k<sub>1</sub>) || bytes(k<sub>2</sub>)''
+* Return ''secnonce'' and ''pubnonce''
+
+==== Nonce Aggregation ====
+
+* The number ''u'' of ''pubnonces'' with ''0 < u < 2^32''
+* The public nonces ''pubnonce<sub>1..u</sub>'': ''u'' 66-byte arrays
+
+The algorithm ''NonceAgg(pubnonce<sub>1..u</sub>)'' is defined as:
+* For ''i = 1 .. 2'':
+** For ''j = 1 .. u'':
+*** Let ''R<sub>i,j</sub> = lift_x(int(pubnonce<sub>j</sub>[i*33:i*33+33]))''; fail if that fails
+** Let ''R'<sub>i</sub> = R<sub>i,1</sub> + R<sub>i,2</sub> + ... + R<sub>i,u</sub>''
+** Let ''R<sub>i</sub> = R'<sub>i</sub>'' if not ''is_infinite(R'<sub>i</sub>)'', otherwise let R<sub>i</sub> = G''
+* Return ''aggnonce = cbytes(R<sub>1</sub>) || cbytes(R<sub>2</sub>)''
+
+Note: If ''is_infinite(R'<sub>i</sub>)'' there is at least one dishonest signer.
+If we would fail here, we will never be able to determine who it is.
+Therefore, we should continue such that the culprit is revealed when collecting and verifying partial signatures.
+However, dealing with the point at infinity requires defining a serialization and may require extra code complexity in implementations.
+Instead, we set the aggregate nonce to some arbitrary point, the generator.
+This is secure, because it only restricts the abilities of the attacker: an attacker that forces the sum of nonces to be infinity by sending some maliciously generated nonce pairs can be turned into an attacker that forces the sum to be the generator (by simply adding the generator to one of the malicious nonces), and this does not change the winning condition of the signature security game (EUF-CMA).
+
+==== Signing ====
+
+Input:
+* The secret nonce ''secnonce'' that has never been used as input to ''Sign'' before: a 64-byte array
+* The secret key ''sk'': a 32-byte array
+* The aggregate public nonce ''aggnonce'': a 66-byte array
+* The number ''u'' of public keys with ''0 < u < 2^32''
+* The public keys ''pk<sub>1..u</sub>'': ''u'' 32-byte arrays
+* The message ''m'': a 32-byte array
+
+The algorithm ''Sign(secnonce, sk, aggnonce, pk<sub>1..u</sub>, m)'' is defined as:
+* Let ''R<sub>1</sub> = lift_x(int(aggnonce[0:33])), R<sub>2</sub> = lift_x(int(aggnonce[33:66]))''; fail if that fails
+* Let ''Q = KeyAgg(pk<sub>1..u</sub>)''; fail if that fails
+* Let ''b = int(hash<sub>MuSig/noncecoef</sub>(aggnonce,bytes(Q),m)) mod n''
```

Comment:
Besides instantiating the random oracle with a hash function, how is it different from the paper?

Reply:
I was just referring to the ordering (paper has X, R, m); probably not a big deal in retrospect

Reply:
Hm I don't think that's noteworthy because there's a clear difference between an abstract description of the scheme (MuSig paper) and a spec. One can imagine that `RO(X, R, m)` is instantiated as `hash(bytes(R),bytes(X),m)`. But generally, if there are other significant differences (for example setting aggnonce to G if infinite otherwise) we should definitely point them out.